src/main/java/util/ApiQuotaManager.java

```diff
@@ -0,0 +1,181 @@
+     *         apiCallsUsedInPreviousRefresh will be returned when newRemainingApiRequests > remainingApiRequests.
+     */
+    private int computeApiCallsUsedInPreviousRefresh(int newRemainingApiRequests) {
+        int apiUsageInRecentRefresh = remainingApiRequests - newRemainingApiRequests;
+
+        if (apiUsageInRecentRefresh < 0) {
+            return apiCallsUsedInPreviousRefresh;
+        }
+
+        return apiUsageInRecentRefresh;
+    }
+
+    /**
+     * Computes the TickerTimer period that is used for refreshing the issues periodically
+     * Assumes future refreshes will take the same number of API calls as the last refresh and finds out the refresh
+     * duration that will spread out the refreshes until the next API quota top up.
+     *
+     * Calculation avoids scheduling a refresh that happens during the same time as the API quota top up.
+     *
+     * PMD warning is suppressed to allow explicit parenthesis.
+     *
+     * @param apiQuota The remaining allowed API requests until the next API quota renewal.
+     *                 Pre-condition: >= 0
+     * @param remainingTimeInMins The remaining time left until the next API quota renewal.
+     *                           Pre-condition: >= 0
+     * @param apiCallsUsedInPreviousRefresh The amount of API used in the previous API refresh.
+     *                                      Pre-condition: >= 0
+     * @param apiQuotaBuffer The amount of API calls that is set aside for manual operations by the user.
+     *                       Pre-condition: >= 0
+     * @param minRefreshPeriod The minimal refresh period that will be used.
+     *                         Recommended value : ApiQuotaManager.DEFAULT_REFRESH_PERIOD_IN_MINS
+     *                         Pre-condition: > 0
+     * @return the computed refresh period or minRefreshPeriod.
+     *          There are 3 conditions that minRefreshPeriod will be used.
+     *          1) During application initialisation, when apiCallsUsedInPreviousRefresh is zero.
+     *          2) When computed refresh rate is < minRefreshPeriod.
+     *          3) When remainingTimeInMins = 0. This indicates that refreshes can occur in the
+     *             nearest time possible.
+     */
+    @SuppressWarnings("PMD")
+    public static long computeRefreshTimerPeriod(int apiQuota, long remainingTimeInMins,
+                                                 int apiCallsUsedInPreviousRefresh,
+                                                 int apiQuotaBuffer, int minRefreshPeriod) {
+
+        assert apiQuota >= 0 && remainingTimeInMins >= 0 && apiCallsUsedInPreviousRefresh >= 0
+                && apiQuotaBuffer >= 0 && minRefreshPeriod > 0;
+
+        int remainingApiQuota = apiQuota - apiQuotaBuffer;
+
+        if ((remainingApiQuota > 0 && apiCallsUsedInPreviousRefresh == 0) || remainingTimeInMins == 0) {
+            return minRefreshPeriod;
+        }
+
+        long refreshTimeInMins;
+        final int bufferTime = 1;
```

Comment:
I think you can get rid of this. It is fine if a refresh is attempted at the same time the quota is topped up. We can afford to have at most one failed refresh per hour. I most cases that refresh will go through anyway because we have a quota buffer. On the other hand, adding a minute to the refresh period every time is more damaging to the user experience.